internal/controller/podpool_controller.go

```diff
@@ -0,0 +1,324 @@
+		Complete(r)
+}
+
+// deleteFinalizer deletes the finalizer from the podPool
+func (r *PodPoolReconciler) deleteFinalizer(ctx context.Context, podPool *kwoksigsv1beta1.PodPool) error {
+	controllerutil.RemoveFinalizer(podPool, controllerFinalizer)
+	return r.Update(ctx, podPool)
+}
+
+// statusConditionController updates the status of the podPool
+func (r *PodPoolReconciler) statusConditionController(ctx context.Context, podPool *kwoksigsv1beta1.PodPool, condition metav1.Condition) error {
+	meta.SetStatusCondition(&podPool.Status.Conditions, condition)
+	return r.Status().Update(ctx, podPool)
+}
+
+// addFinalizer adds the finalizer to the podPool
+func (r *PodPoolReconciler) addFinalizer(ctx context.Context, podPool *kwoksigsv1beta1.PodPool) error {
+	controllerutil.AddFinalizer(podPool, controllerFinalizer)
+	return r.Update(ctx, podPool)
+}
+
+// get pods in the cluster with owner reference to the podPool
+func (r *PodPoolReconciler) getPods(ctx context.Context, podPool *kwoksigsv1beta1.PodPool) ([]corev1.Pod, error) {
+	pods := &corev1.PodList{}
+	err := r.List(ctx, pods, client.InNamespace(podPool.Namespace), client.MatchingLabels{controllerLabel: podPool.Name})
+	if err != nil && strings.Contains(err.Error(), "does not exist") {
+		return []corev1.Pod{}, nil
+	} else if err != nil {
+		return nil, err
+	}
+	return pods.Items, nil
+}
+
+// deletePods deletes the pods in the cluster
+func (r *PodPoolReconciler) deletePods(ctx context.Context, podPool *kwoksigsv1beta1.PodPool, pods []corev1.Pod) error {
+	for i := int32(len(pods)); i > podPool.Spec.PodCount; i-- {
+		// Delete a pod
+		err := r.Delete(ctx, &pods[i-1])
+		if err != nil {
+			return err
+		}
+	}
+	err := r.updateObservedGeneration(ctx, podPool)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// updateObservedGeneration updates the observed generation of the podPool
+func (r *PodPoolReconciler) updateObservedGeneration(ctx context.Context, podPool *kwoksigsv1beta1.PodPool) error {
+	podPool.Status.ObservedGeneration = podPool.Generation
+	return r.Status().Update(ctx, podPool)
+}
+
+func (r *PodPoolReconciler) createPods(ctx context.Context, podPool *kwoksigsv1beta1.PodPool, pods []corev1.Pod) error {
+	podLabels := podPool.Spec.PodTemplate.Labels
+	if podLabels == nil {
+		podLabels = make(map[string]string)
+	}
+	podLabels[controllerLabel] = podPool.Name
+	podToleration := podPool.Spec.PodTemplate.Spec.Tolerations
+	if podToleration == nil {
+		podToleration = make([]corev1.Toleration, 0)
+	}
+	podToleration = append(podToleration, corev1.Toleration{
+		Key:      controllerAnnotation,
+		Operator: corev1.TolerationOpExists,
+		Effect:   corev1.TaintEffectNoSchedule,
+	})
+
+	podAnnotation := podPool.Spec.PodTemplate.Annotations
+	if podAnnotation == nil {
+		podAnnotation = make(map[string]string)
+	}
+	podAnnotation[controllerAnnotation] = fakeString
+	for i := int32(len(pods)); i < podPool.Spec.PodCount; i++ {
+		// Create a new pod
+		pod := &corev1.Pod{
```

Comment:
Why creating the pod object inside the loop if you are not using `i` ?

Reply:
each one of them getting other name ... i'm generating the name of the pod with the name of the CRD